Add flag to optionally listen on all interfaces

keilin-anz opened the pull request:
Server currently listens on `localhost:${port}` which is fine for most scenarios
This adds the ability to optionally listen on `:${port}` instead (all local IPs)
It's unknown whether this has far reaching implications - very much included just as a jump off point for samuong/alpaca#95 if desired

Potentially addresses samuong/alpaca#95

**NB: previous commits removed since it was a tiny change and clashed with recent merges**

dylanANZ commented:
I was thinking possibly `-l` and `--listen` which set an address, say `0:0:0:0` or a regular IP.
Default value is still `localhost`

keilin-anz commented:
> I was thinking possibly -l or --listen which set an address, say 0.0.0.0 or a regular IP.
> Default value is still localhost

Looks like @samuong is on board with that, will update

Amusingly, I was intentionally avoiding that implementation in some misguided attempt to make listening on `:3128` easy - but the proposed approach is definitely much more flexible

samuong commented on main.go:
```diff
@@ -36,6 +36,7 @@ func whoAmI() string {
 
 func main() {
 	log.SetFlags(log.LstdFlags | log.Lshortfile)
+	location := flag.String("l", "localhost", "IPv4 address to listen on")
```
This isn't just going to accept an "ipv4 address" - it'll accept ipv6 addresses and hostnames too? I'm thinking we should say something like "address to listen on" instead

samuong replied:
Actually this isn't that simple, IPv6 addresses need to be enclosed in square brackets before being passed to [net.Dial](https://pkg.go.dev/net#Dial)...

samuong commented on main.go:
```diff
@@ -91,7 +92,7 @@ func main() {
 	s := &http.Server{
 		// Set the addr to localhost so that we only listen locally.
-		Addr:    fmt.Sprintf("localhost:%d", *port),
+		Addr:    fmt.Sprintf("%s:%d", *location, *port),
```
...so here, we should probably do something like:

```
Addr: net.JoinHostPort(*location, strconv(*port)),
```

That'll make sure that it gets enclosed in square brackets, if it's an IPv6 address.

keilin-anz commented:
Resolving rebase conflicts from `main` was pointless, so dropped old commits and redid addressing PR comments

(very lightly) tested with default value (`localhost:3128`), empty value (`:3128`), and an arbitrary local IPv6 address

samuong reviewed (approved):
lgtm, thanks @keilin-anz!
